book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+### Summary
+
+- **Sled reads** may be done synchronously (in `call`) or asynchronously (in
+  the `Future`), depending on the context;
+
+- **Sled writes** must be done synchronously (in `call`)
+
+## In-memory data structures
+[in-memory]: #in-memory
+
+At a high level, the in-memory data structures store a collection of chains,
+each rooted at the highest finalized block. Each chain consists of a map from
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+### The `Chain` type
+[chain-type]: #chain-type
+
+
+The `Chain` type represents a chain of blocks. Each block represents an
+incremental state update, and the `Chain` type caches the cumulative state
+update from its root to its tip.
+
+The `Chain` type is used to represent the non-finalized portion of a complete
+chain of blocks rooted at the genesis block. The parent block of the root of
+a `Chain` is the tip of the finalized portion of the chain.
+
+The `Chain` type supports serveral operations to manipulate chains, `push`,
+`pop_root`, and `fork`. `push` is the most fundamental operation and handles
+contextual validation of chains as they are extended. `pop_root` is provided
+for finalization, and is how we move blocks from the non-finalized portion of
+the state to the finalized portion. `fork` on the other hand handles creating
+new chains for `push` when new blocks arrive whose parent isn't a tip of an
+existing chain.
+
+The `Chain` type is defined by the following struct and API:
+
+```rust
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, tx_index)>,
+
+    utxos: HashSet<transparent::Output>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cumulative data members
```

Comment:
Just some quick thoughts on base and edge cases we need to look out for:
* genesis - empty finalized and empty non-finalized state
* restart - empty non-finalized state
* fork at finalized tip - pop entire chain / fork empty Chain
* checkpoint to block verifier transition - start of contextual validation

Are there any others?

Reply:
Subtract work

Pathological case - finalising a very long side-chain

Reply:
- [ ] duplicate blocks

Reply:
lgtm